tests/tensor/random/rewriting/test_basic.py

```diff
@@ -566,10 +629,110 @@ def test_DimShuffle_lift(ds_order, lifted, dist_op, dist_params, size, rtol):
             ),
             (2, 3),
         ),
-        # Multivariate cases, indexing only supported if it does not affect core dimensions
         (
-            # Indexing dips into core dimension
-            (np.array([1]), 0),
+            # Empty-slice
+            (slice(None), slice(10, None), slice(1, None)),
+            True,
+            normal,
+            (
+                np.arange(30).reshape(2, 3, 5),
+                np.full((1, 5), 1e-6),
+            ),
+            (2, 3, 5),
+        ),
+        (
+            # Multidimensional boolean indexing
+            (rand_bool_mask((5, 3, 2)),),
+            True,
+            normal,
+            (
+                np.arange(30).reshape(5, 3, 2),
+                1e-6,
+            ),
+            (),
+        ),
+        (
+            # Multidimensional boolean indexing
+            (rand_bool_mask((5, 3)),),
+            True,
+            normal,
+            (
+                np.arange(30).reshape(5, 3, 2),
+                1e-6,
+            ),
+            (),
+        ),
+        (
+            # Multidimensional boolean indexing
+            (rand_bool_mask((5, 3)), slice(None)),
+            True,
+            normal,
+            (
+                np.arange(30).reshape(5, 3, 2),
+                1e-6,
+            ),
+            (),
+        ),
+        # 20
+        (
+            # Multidimensional boolean indexing
+            (slice(None), rand_bool_mask((3, 2))),
+            True,
+            normal,
+            (
+                np.arange(30).reshape(5, 3, 2),
+                1e-6,
+            ),
+            (),
+        ),
+        (
+            # Multidimensional boolean indexing
+            (rand_bool_mask((5, 3)),),
+            True,
+            normal,
+            (
+                np.arange(3).reshape(1, 3, 1),
+                np.full((5, 1, 2), 1e-6),
+            ),
+            (5, 3, 2),
+        ),
+        (
+            # Multidimensional boolean indexing
+            (
+                np.array([True, False, True, False, False]),
+                slice(None),
+                (np.array([True, True])),
+            ),
+            True,
+            normal,
+            (
+                np.arange(30).reshape(5, 3, 2),
+                1e-6,
+            ),
+            (),
+        ),
+        pytest.param(
+            # Multidimensional boolean indexing
+            (
+                np.array([True, False, True, False, False]),
+                slice(None),
+                (np.array([True, False])),
+            ),
+            True,
+            normal,
+            (
+                np.arange(30).reshape(5, 3, 2),
+                1e-6,
+            ),
+            (),
+            marks=pytest.mark.xfail(
```

Comment:
@aseyboldt this test failed because of the changes to `broadcast_shape_iter`. The intermediate shape inference tries to broadcast the two boolean.nonzero() entries which have shape (2,) and (1,). 

Because those are created inside the rewrite there is no chance for constant_folding to take place before it's too late (or for the user to put `specify_shapes` in the right places.

We could constant_fold in the rewrite, but for now I decided to leave as is. Just mentioning to add to our list of the edge cases of runtime-broadcasting.

In theory there may be nothing wrong with allowing runtime broadcasting of indexes or shape arguments that are not direct outputs of anything.

I considered adding a `allow_runtime_broadcast` flag to `broadcast_shape_iter`, but maybe that's better left for another PR.